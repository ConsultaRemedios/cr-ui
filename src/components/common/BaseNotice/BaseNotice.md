### Basic Usage

The BaseNotice works fully dinamically, the `title`, `message` and the `icons`. For `title` and `message` you should send simple strings. For
Icon, it works the same as `BaseIcon` component.
The component also has a `timeout` prop, it works with miliseconds, default is 5000, if you need more, just send as prop as we use in examples:

```
const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`) ;
<BaseNotice
  title="Something wrong happened"
  message="Lorem ipsum dolor sit amet"
/>
```
#### Types
```
const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`) ;

<div>
  <BaseNotice
    title="Something wrong happened"
    message="Lorem ipsum dolor!!"
    type="danger"
    :icon="getIcon(icons, 'warning')"
  />

  <br />

  <BaseNotice
    title="Sorry, you got an error!"
    message="Lorem ipsum dolor sit amet"
    type="warning"
    :icon="getIcon(icons, 'info')"
  />

  <br />

  <BaseNotice
    title="Congrats all works!"
    type="success"
    :icon="getIcon(icons, 'thumb-up')"
  />

  <br />

  <BaseNotice
    title="Something wrong happened"
    message="Lorem ipsum dolor?"
    type="info"
    :icon="getIcon(icons, 'help')"
  />


</div>
```
#### Snackbar

Defines the Notice type as snackbar, it should be `true` or `false`, default is `false`.
You should also use a method to control the Fading, see the example:

```js
new Vue({
  data() {
    return {
      error: '',
    }
  },
  methods: {
    setError() {
      this.error = 'Houve um erro, vocÃª precisa usar uma outra senha';
    },

    clearError() {
      this.error = '';
    },
  },

  template: `
    <div>
      <GridCol :span="{ md: 6 }" :offset="{ md: 4 }">
        <BaseNotice
          v-if="error"
          :snackbar="true"
          :title="error"
          type="danger"
          :scroll-to-self="false"
          :auto-close="true"
          @hide="clearError"
        />
      </GridCol>

      <BaseButton @click="setError">Mostrar Snackbar</BaseButton>
    </div>
  `
})
```
