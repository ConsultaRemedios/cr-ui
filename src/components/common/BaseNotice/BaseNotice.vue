<template>
<transition
  :enter-class="$style.enter"
  :enter-active-class="$style.enterActive"
  :enter-to-class="$style.enterTo"
  :leave-class="$style.leave"
  :leave-active-class="$style.leaveActive"
  :leave-to-class="$style.leaveTo"
  @after-leave="$emit('hide')"
>
  <div
    v-show="show"
    :class="[{[$style.snackbar]: this.snackbar} , $style.notice, $style[type]]"
  >
    <ContainerBlock>
      <div :class="$style.content">
        <div :class="$style.iconBlock">
          <BaseIcon :id="icon.id" />
        </div>

       <div :class="$style.textBlock">
         <span :class="$style.title">
          {{title}}
          </span>

          <p v-if="message" :class="$style.message">
            {{message}}
          </p>
       </div>
      </div>
    </ContainerBlock>
  </div>
</transition>
</template>

<script>
  import ContainerBlock from './../ContainerBlock';
  import infoIcon from './../../../icons/info.icon.svg';
  import BaseIcon from '../BaseIcon';

  export default {
    name: 'BaseNotice',
    components: {
      ContainerBlock,
      BaseIcon,
    },

    data() {
      return {
        show: false,
      };
    },

    props: {
      type: {
        type: String,
        default: 'info',
        validator(value) {
          return ['info', 'warning', 'success', 'danger'].includes(value);
        },
      },

      title: {
        type: String,
        required: true,
      },

      message: {
        type: String,
      },

      scrollToSelf: {
        type: Boolean,
        default: true,
      },

      icon: {
        type: Object,
        default() { return infoIcon; },
      },

      timeout: {
        type: Number,
        default: 5000,
      },

      snackbar: {
        type: Boolean,
        default: false,
      },

      autoClose: {
        type: Boolean,
        default: false,
      },
    },

    mounted() {
      if (this.scrollToSelf && !this.snackbar) {
        window.scrollTo(0, this.$el.offsetTop);
      }

      this.show = true;

      if (this.autoClose) {
        setTimeout(() => {
          this.show = false;
        }, this.timeout);
      }
    },
  };
</script>

<style module>
  @value media-md, media-lg from './../../../styles/variables.css';

  .content {
    display: inline-flex;
    align-items: center;
  }

  .iconBlock {
      padding-right: 10px;
    }

  .notice {
    color: #fff;
    padding: 15px 0;
  }

  .icon {
    display: none;
  }

  .title {
    font-weight: 800;
  }

  .message {
    margin: 0;
  }

  .info {
    background: #aaaaaa;
  }

  .warning {
    background: #fb9c12;
  }

  .success {
    background: #1BB738
  }

  .danger {
    background: #D30B3F;
  }

  .snackbar {
    min-width: 300px;
    border-radius: 5px;
    z-index: 99999;
    position: fixed;
    top: 10px;
    left: 10px;
    right: 10px;
    box-shadow: 0 3px 4px 0 rgba(0,0,0,0.34);
  }

  .snackbar.info {
    background: rgb(170, 170, 170, 0.94);
  }

  .snackbar.warning {
    background: rgb(251, 156, 18, 0.94);
  }

  .snackbar.success {
    background: rgb(27, 183, 56, 0.94);
  }

  .snackbar.danger {
    background: rgba(211, 11, 63, 0.94);
  }

  .snackbar.enterActive,
  .snackbar.leaveActive {
    transition: opacity 0.2s ease-out,
                transform 0.2s ease-out;
  }

  .snackbar.enter,
  .snackbar.leaveTo {
    opacity: 0;
    transform: translateY(-100%);
  }

  .snackbar.enterTo,
  .snackbar.leave {
    opacity: 1;
    transform: translateY(0);
  }

  @media media-md {
    .snackbar {
      top: 20px;
      left: 50%;
      right: auto;
      transform: translateX(-50%);
      min-width: 680px;
    }

    .snackbar.enter,
    .snackbar.leaveTo {
      transform: translateX(-50%) translateY(-100%);
    }

    .snackbar.enterTo,
    .snackbar.leave {
      transform: translateX(-50%) translateY(0);
    }
  }

  @media media-lg {
    .snackbar {
      top: 95px;
      padding: 15px 10px;
    }

    .content {
      position: relative;
      align-items: center;
    }

    .iconBlock {
      padding-right: 17px;
    }

    .icon {
      display: block;
      position: absolute;
      top: 50%;
      left: 0;
      transform: translateY(-50%);
    }
  }

</style>
