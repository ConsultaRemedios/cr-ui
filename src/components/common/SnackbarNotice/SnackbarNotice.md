We have many different variations on BaseNotice component. The **SnackbarNotice** encapsulates
the BaseNotice component with specific prop values into it in order to increase the UI consistency
for user feedback messages

### Types

```js
let showError = false;
let showSuccess = false;

<div>
  <SnackbarNotice v-if="showError"
    type="error"
    content="Ops, something wrong happened"
    @hide="showError = false"
  />

  <SnackbarNotice v-if="showSuccess"
    type="success"
    content="Everything is fine"
    @hide="showSuccess = false"
  />

  <BaseButton @click="showError = true">
    Show error SnackbarNotice
  </BaseButton>

  <BaseButton @click="showSuccess = true">
    Show success SnackbarNotice
  </BaseButton>
</div>
```

### Content prop as a string

```js
let error = false;

<div>
  <SnackbarNotice v-if="error"
    type="error"
    content="Ops, something wrong happened"
    @hide="error = false"
  />

  <BaseButton @click="error = true">
    Show SnackbarNotice with content as string
  </BaseButton>
</div>
```

### Content prop as an object

```js
let error = false;

<div>
 <SnackbarNotice v-if="error"
    type="error"
    :content="{ title: 'Ops, something wrong happened', message: 'Something more detailed here' }"
    @hide="error = false"
  />

  <BaseButton @click="error = true">
    Show SnackbarNotice with content as object
  </BaseButton>
</div>
```
