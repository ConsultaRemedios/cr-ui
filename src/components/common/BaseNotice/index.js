import BaseNotice from './BaseNotice.vue';

export default BaseNotice;
