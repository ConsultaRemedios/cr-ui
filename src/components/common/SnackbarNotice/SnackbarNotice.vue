<template>
  <BaseNotice
    v-bind="props"
    :type="baseNoticeType"
    :timeout="timeout"
    :snackbar="true"
    :auto-close="true"
    @hide="$emit('hide')"
  />
</template>

<script>
import BaseNotice from '../BaseNotice';
import thumbsUpIcon from '../../../icons/thumb-up.icon.svg';
import infoIcon from '../../../icons/info.icon.svg';

export default {
  name: 'SnackbarNotice',
  components: { BaseNotice },
  props: {
    type: {
      type: String,
      required: true,
      validator(value) {
        return ['success', 'error', 'warning'].includes(value);
      },
    },

    content: {
      type: [String, Object],
      required: true,
      validator(value) {
        if (typeof value === 'string') return true;

        const keys = Object.keys(value);
        return keys.includes('title') && keys.includes('message');
      },
    },

    timeout: {
      type: Number,
      default: 5000,
    },
  },

  computed: {
    props() {
      if (typeof this.content === 'string') {
        return { title: this.content };
      }

      const { title, message } = this.content;
      return { title, message };
    },

    icon() {
      return this.type === 'success' ? thumbsUpIcon : infoIcon;
    },

    baseNoticeType() {
      if (this.type === 'error') return 'danger';
      return this.type;
    },
  },
};
</script>
