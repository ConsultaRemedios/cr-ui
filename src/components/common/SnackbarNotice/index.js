import SnackbarNotice from './SnackbarNotice.vue';

export default SnackbarNotice;
