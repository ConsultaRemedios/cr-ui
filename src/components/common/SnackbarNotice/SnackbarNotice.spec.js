import { shallowMount } from '@vue/test-utils';
import SnackbarNotice from './SnackbarNotice.vue';
import BaseNotice from './../BaseNotice';

describe('SnackbarNotice', () => {
  let wrapper;

  const setup = (props = {}) => {
    const defaultProps = {
      content: 'A content message',
      type: 'success'
    };

    const propsData = { ...defaultProps, ...props };
    return shallowMount(SnackbarNotice, { propsData });
  };

  beforeEach(() => {
    wrapper = null;
  });

  describe('Props', () => {
    describe('#type', () => {
      it('is valid when type is equal to success', () => {
        wrapper = setup();
        const { validator } = wrapper.vm.$options.props.type;
        expect(validator('success')).toBeTruthy();
      });

      it('is valid when type is equal to error', () => {
        wrapper = setup();
        const { validator } = wrapper.vm.$options.props.type;
        expect(validator('error')).toBeTruthy();
      });

      it('is valid when type is equal to warning', () => {
        wrapper = setup();
        const { validator } = wrapper.vm.$options.props.type;
        expect(validator('warning')).toBeTruthy();
      });

      it('is invalid when type is different than success or error', () => {
        wrapper = setup();
        const { validator } = wrapper.vm.$options.props.type;
        expect(validator('foobar')).toBeFalsy();
      });
    });

    describe('#content', () => {
      it('is invalid when it is an object without the required keys', () => {
        wrapper = setup();

        const { validator } = wrapper.vm.$options.props.content;

        expect(validator({ title: 'Some title' })).toBeFalsy();
        expect(validator({ message: 'Some title' })).toBeFalsy();
      });

      it('is valid when it is an object with the required keys', () => {
        wrapper = setup();

        const { validator } = wrapper.vm.$options.props.content;

        expect(validator({
          title: 'Some title',
          message: 'Some message example here'
        })).toBeTruthy();
      });
    });

    describe('#timeout', () => {
      it('returns 5000 when the prop is not passed', () => {
        wrapper = setup();
        expect(wrapper.vm.timeout).toEqual(5000);
      });
    });
  });

  describe('Computed', () => {
    describe('#props', () => {
      it('returns the proper plain object if "content" prop is a string', () => {
        wrapper = setup();

        expect(wrapper.vm.props).toEqual({
          title: 'A content message',
        });
      });

      it('returns the proper plain object if "content" is a valid object', () => {
        wrapper = setup({
          content: {
            title: 'Some title',
            message: 'Some message here'
          }
        });

        expect(wrapper.vm.props).toEqual({
          title: 'Some title',
          message: 'Some message here',
        });
      });
    });

    describe('#icon', () => {
      it('returns info-icon if type is equal to error', () => {
        wrapper = setup({
          type: 'error',
          content: 'Some title'
        });

        expect(wrapper.vm.icon).toEqual({ id: 'info.icon.svg', viewbox: '0 0 24 24' });
      });

      it('returns thumb-up icon if type is not equal to error (success)', () => {
        wrapper = setup({
          type: 'success',
          content: 'Some title'
        });

        expect(wrapper.vm.icon).toEqual({ id: 'thumb-up.icon.svg', viewbox: '0 0 24 24' });
      });
    });

    describe('#baseNoticeType', () => {
      it('returns danger when type is equal to error', () => {
        wrapper = setup({ type: 'error' });
        expect(wrapper.vm.baseNoticeType).toEqual('danger');
      });

      it('returns success when type is not equal to error', () => {
        wrapper = setup({ type: 'success' });
        expect(wrapper.vm.baseNoticeType).toEqual('success');
      });

      it('returns warning when type is equal to warning', () => {
        wrapper = setup({ type: 'warning' });
        expect(wrapper.vm.baseNoticeType).toEqual('warning');
      });
    });
  });

  describe('Render', () => {
    const props = {
      icon: { id: 'info.icon.svg', viewbox: '0 0 24 24' },
      snackbar: true,
      autoClose: true,
      scrollToSelf: true,
      type: 'success'
    };

    describe('Type is error', () => {
      it('renders BaseNotice properly', () => {
        wrapper = setup({
          content: 'The title here',
          type: 'error'
        });

        expect(wrapper.find(BaseNotice).vm.$props).toEqual({
          ...props,
          type: 'danger',
          title: 'The title here',
          message: undefined,
          timeout: 5000,
        });
      });
    });

    describe('Type is success', () => {
      it('renders BaseNotice properly', () => {
        wrapper = setup({
          content: 'The title here',
          type: 'success'
        });

        expect(wrapper.find(BaseNotice).vm.$props).toEqual({
          ...props,
          type: 'success',
          title: 'The title here',
          message: undefined,
          timeout: 5000,
        });
      });
    });

    describe('Content as string', () => {
      it('renders BaseNotice properly', () => {
        wrapper = setup({
          content: 'The title here',
          timeout: 6000
        });

        expect(wrapper.find(BaseNotice).vm.$props).toEqual({
          ...props,
          title: 'The title here',
          message: undefined,
          timeout: 6000,
        });
      });
    });

    describe('Content as valid object', () => {
      it('render BaseNotice properly', () => {
        wrapper = setup({
          content: {
            title: 'The important title here',
            message: 'An error details here if you want to',
          },
          timeout: 3000
        });

        expect(wrapper.find(BaseNotice).vm.$props).toEqual({
          ...props,
          title: 'The important title here',
          message: 'An error details here if you want to',
          timeout: 3000,
        });
      });
    });

    it('Matches the snapshot', () => {
      wrapper = setup({
        content: {
          title: 'The important title here',
          message: 'An error details here if you want to',
        },
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Behavior', () => {
    it('emits "hide" event when BaseNotice emits hide event', () => {
      wrapper = setup();

      jest.spyOn(wrapper.vm, '$emit');

      wrapper.find(BaseNotice).vm.$emit('hide');

      expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
      expect(wrapper.vm.$emit).toHaveBeenCalledWith('hide');
    });
  });
});
