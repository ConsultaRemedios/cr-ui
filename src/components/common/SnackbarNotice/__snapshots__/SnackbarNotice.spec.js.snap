// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarNotice Render Matches the snapshot 1`] = `
<basenotice-stub
  autoclose="true"
  icon="[object Object]"
  message="An error details here if you want to"
  scrolltoself="true"
  snackbar="true"
  timeout="5000"
  title="The important title here"
  type="success"
/>
`;
